src/main/java/org/truffleruby/language/constants/GetConstantNode.java

```diff
@@ -71,6 +75,16 @@ protected Object autoloadConstant(LexicalScope lexicalScope, DynamicObject modul
             return doMissingConstant(module, name, getSymbol(name));
         }
 
+        String featureString = StringOperations.getString(feature);
+        final String expandedPath = getContext().getFeatureLoader().findFeature(featureString);
+        if (expandedPath != null) {
+            final ReentrantLockFreeingMap<String> fileLocks = getContext().getFeatureLoader().getFileLocks();
+            final ReentrantLock lock = fileLocks.get(expandedPath);
+            if (lock.isHeldByCurrentThread()) {
+                return null;
+            }
+        }
```

Comment:
I removed some fairly similar code in 6bd14d83c52fcb9108811c5905ffdf5780214731 but that was probably a mistake.
I'm not sure why this is needed and the above `autoloadConstant.getAutoloadConstant().isAutoloadingThread()` is not enough, but it clearly seems to result in better behavior.
Thanks for the PR!

Reply:
@XrXr did all the investigation, so I'll leave to him to answer why that is not enough if he knows.

Reply:
It's not enough because `require`ing a file doesn't set all the autoload constants related to that file on the heap into autoload mode. This is why the above check falls through in the added test.

Reply:
Indeed, the `registeredAutoloads` map only keeps one entry per `path`.
Maybe we should have it as `Map<String basename, Map<String autoload path, RubyConstant[] constants>>`, do you know if MRI does something similar?

In any case I think I'll merge this fix first as it looks good to me.